### ISSUE TITLE
realmd: The modal to Join a domain shows an empty notification on open

### ISSUE THREAD

**Issue by KKoukiou** [cockpit-project/cockpit#13157]
Cockpit version: 207
OS: systemd/realmd

With the recent notifcation porting to PF4 this bug got introduced where we notifications are shown even if on the html the have hidden attribute.

Interestingly elements which are set with diplay: grid will not be hidden if passed the hidden attribute.

I noticed that on the modal to Join a Domain but I won't be surprised if this is introduced somewhere else as well.

@garrett what's the proper way to hide a div element that also has display: grid on it?

![Screen Shot 2019-11-15 at 17 49 56](https://user-images.githubusercontent.com/14921356/68960396-60fce580-07d0-11ea-8a49-838064650ddd.png)

**Comment by garrett** [comment]
You can't have a `display: none` and `display: grid`. It has to be one or the other.

Related: https://css-tricks.com/the-hidden-attribute-is-visibly-weak/

This bit of CSS should enforce the hidden attribute (given that the element isn't more specific and/or comes later to `!important` the `display` attribute again):

```css
[hidden] {
  display: none !important;
}
```

But, the screenshot makes it look like the element in question is actually a child of a grid element? Then it's something entirely different. (I'll go look at the source now.)

**Comment by garrett** [comment]
OK, so this is a grid. I wonder if this is actually a bug in PF4 then?

**Comment by garrett** [comment]
As hidden is a boolean, we'd want something like:

```css
[hidden]:not([hidden=false]) {
  display: none !important;
}
```

Hidden can also have no value, which then becomes `hidden="hidden"`, which basically equates to hidden=true, hence `[hidden]` without a value in the CSS.

**Comment by garrett** [comment]
Cockpit's `<body>` tag for the shell itself uses an inline `display: block` to override hidden, which is (incorrectly) set to `hidden=""`. It looks like this in the DOM (at least in Firefox's developer tools):

```html
<body style="display: block;" hidden="">
```

This _should_ just simply be:

```html
<body hidden="false">
```

or simpler still, the attribute could be removed entirely:

```html
<body>
```

PatternFly (3? I think?) has this snippet of CSS in `patternfly-base.css` (which gets compiled into (`patternfly.css`):

```css
[hidden], template {
  display: none;
}
```

I've tried the following CSS:

```css
[hidden=true], [hidden=hidden] {
  display: none !important;
}
```

But it doesn't work for what we want either, as it doesn't actually handle the `<div hidden>` case now. (And when I try to get it to do so, it hides the page again. Or I try `[hidden=]` and it even fails to compile due to CSS checks.)

In other words, I don't think the fix can be general and in CSS. We'll need to fix the HTML (in multiple places, to fix things properly).

**Comment by garrett** [comment]
I've looked through our codebase and see the `hidden` attribute being used in a lot of HTML. When something is unhidden, often with jQuery's `.show()` (opposite of `.hide()`), the code doesn't usually remove the hidden attribute. This relies on the browser to use CSS to override the hidden attribute to show something.

If we override it with the above CSS, _then we're going to break a lot of Cockpit_.

Thankfully, browsers — including screen readers paired with browsers — know what to do and show (or speak) unhidden elements (even when they have a hidden attribute).

My suggestion:

1. Fix this one instance of the empty notification showing.
2. Fix any other notifications not showing (if any).
3. Later on, consider fixing Cockpit's usage of hidden — but this should be way, way later. Also, it will _probably_ be _mostly_ fixed the move code we move away from jQuery and to React.
